pyfr/integrators/dual/pseudosteppers.py

```diff
@@ -13,32 +13,27 @@ def collect_stats(self, stats):
         # Total number of pseudo-steps
         stats.set('solver-time-integrator', 'npseudosteps', self.npseudosteps)
 
-    def _add_with_dts(self, *args, c):
-        vals, regs = list(args[::2]), list(args[1::2])
+    def _rhs_with_dts(self, t, *args, c):
+        # Compute -∇·f
+        self.system.rhs(t, *args)
 
         # Coefficients for the dual-time source term
         svals = [c*sc for sc in self._dual_time_source]
 
-        # Normal addition
-        axnpby = self._get_axnpby_kerns(len(vals))
-        self._prepare_reg_banks(*regs)
-        self._queue % axnpby(*vals)
-
-        # Source addition
+        # Source addition -∇·f - dQ/dt
         axnpby = self._get_axnpby_kerns(len(svals) + 1, subdims=self._subdims)
-        self._prepare_reg_banks(regs[0], self._idxcurr, *self._source_regidx)
+        self._prepare_reg_banks(args[1], self.idxcurr, *self._source_regidx)
         self._queue % axnpby(1, *svals)
 
     def finalise_step(self, currsoln):
-        add = self._add
-        pnreg = self._pseudo_stepper_nregs
+        pnreg, dtsnreg = self._pseudo_stepper_nregs, len(self._dual_time_source)
 
         # Rotate the source registers to the right by one
-        self._regidx[pnreg:] = (self._source_regidx[-1:] +
-                                self._source_regidx[:-1])
+        self._regidx[pnreg:pnreg + dtsnreg - 1] = (self._source_regidx[-1:]
+                                                   + self._source_regidx[:-1])
 
         # Copy the current soln into the first source register
-        add(0.0, self._regidx[pnreg], 1.0, currsoln)
+        self._add(0, self._regidx[pnreg], 1, currsoln)
```

Comment:
In pseudosteppers we have dropped the decimals, so I think it is more consistent like this.